File: samples/python/resilience/dead_letter_agent.py
"""
Dead Letter Agent Pattern
Failed agent tasks are routed to a dead letter handler.
"""
from typing import TypedDict, Literal
from langgraph.graph import StateGraph, END
from langchain_anthropic import ChatAnthropic

llm = ChatAnthropic(model="claude-haiku-4-5-20251001")


class AgentState(TypedDict):
    task: str
    result: str
    error: str
    retries: int
    status: str


def agent_node(state: AgentState) -> AgentState:
    retries = state.get("retries", 0)
    print(f"[agent_node] Attempt #{retries + 1} for task: '{state['task']}'")

    if retries < 1:
        print("[agent_node] Simulating failure on first attempt.")
        return {**state, "error": "Simulated transient failure", "retries": retries + 1, "status": "failed"}

    response = llm.invoke(f"Complete this task in one sentence: {state['task']}")
    result = response.content if hasattr(response, "content") else str(response)
    print(f"[agent_node] Task completed: {result}")
    return {**state, "result": result, "retries": retries + 1, "status": "done"}


def dead_letter_node(state: AgentState) -> AgentState:
    print(f"[dead_letter_node] Task routed to dead letter queue.")
    print(f"[dead_letter_node] Failed task: '{state['task']}'")
    print(f"[dead_letter_node] Error: {state['error']}")
    notification = (
        f"DEAD LETTER NOTIFICATION: Task '{state['task']}' "
        f"failed after {state['retries']} attempt(s). Error: {state['error']}"
    )
    print(f"[dead_letter_node] {notification}")
    return {**state, "result": notification, "status": "dead_letter"}


def route(state: AgentState) -> Literal["dead_letter_node", "agent_node", "__end__"]:
    if state["status"] == "done":
        return END
    if state["status"] == "failed" and state["retries"] >= 2:
        return "dead_letter_node"
    if state["status"] == "failed":
        return "agent_node"
    return END


def build_graph() -> StateGraph:
    graph = StateGraph(AgentState)
    graph.add_node("agent_node", agent_node)
    graph.add_node("dead_letter_node", dead_letter_node)
    graph.set_entry_point("agent_node")
    graph.add_conditional_edges("agent_node", route)
    graph.add_edge("dead_letter_node", END)
    return graph.compile()


def main():
    app = build_graph()

    print("=== Run 1: Task succeeds on retry ===")
    state1: AgentState = {"task": "Summarize the benefits of resilience patterns", "result": "", "error": "", "retries": 0, "status": "pending"}
    output1 = app.invoke(state1)
    print(f"Final status: {output1['status']}\n")

    print("=== Run 2: Task fails and goes to dead letter (force 2 failures) ===")
    state2: AgentState = {"task": "A task that will fail", "result": "", "error": "Simulated transient failure", "retries": 1, "status": "failed"}
    output2 = app.invoke(state2)
    print(f"Final status: {output2['status']}")
    print(f"Dead letter result: {output2['result']}")


if __name__ == "__main__":
    main()

File: samples/typescript/resilience/dead_letter_agent.ts
import { ChatAnthropic } from "@langchain/anthropic";
import { StateGraph, END, START, Annotation } from "@langchain/langgraph";
const llm = new ChatAnthropic({ model: "claude-haiku-4-5-20251001" });
const State = Annotation.Root({
  task: Annotation<string>(),
  result: Annotation<string>(),
  error: Annotation<string>(),
  retries: Annotation<number>(),
  status: Annotation<string>(),
});
type S = typeof State.State;
const agentNode = async (s: S): Promise<Partial<S>> => {
  if (s.retries < 1) return { error: "Simulated failure", status: "failed", retries: s.retries + 1 };
  const r = await llm.invoke(`Complete this task: ${s.task}`);
  return { result: r.content as string, status: "done" };
};
const deadLetterNode = async (s: S): Promise<Partial<S>> => {
  console.log(`[DeadLetter] Task failed: ${s.error}`);
  return { status: "dead_letter" };
};
const route = (s: S) => s.status === "done" ? END : s.status === "failed" && s.retries >= 2 ? "deadLetter" : "agent";
const graph = new StateGraph(State)
  .addNode("agent", agentNode).addNode("deadLetter", deadLetterNode)
  .addEdge(START, "agent").addConditionalEdges("agent", route, { [END]: END, agent: "agent", deadLetter: "deadLetter" })
  .addEdge("deadLetter", END).compile();
async function main() {
  const r = await graph.invoke({ task: "Explain circuit breakers", result: "", error: "", retries: 0, status: "pending" });
  console.log("Status:", r.status, "| Result:", (r.result || r.error).substring(0, 80));
}
main().catch(console.error);
